src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -261,11 +261,11 @@ public void postTenantpermissions(OkapiPermissionSet entity, Map<String, String>
     CQLWrapper all = new CQLWrapper();
     return connection.get(PermsAPI.TABLE_NAME_PERMSUSERS, PermissionUser.class, all, false)
         .compose(results -> {
-          Future<Void> future = Future.succeededFuture();
+          List<Future<Void>> futures = new ArrayList<>(results.getResults().size());
           for (PermissionUser permUser : results.getResults()) {
-            future = future.compose(x -> migratePermsAssignUser(permUser, connection, vertxContext, tenantId));
+            futures.add(migratePermsAssignUser(permUser, connection, vertxContext, tenantId));
           }
-          return future;
+          return GenericCompositeFuture.all(futures).mapEmpty();
```

Comment:
When running in parallel there can be a race condition in PermsAPI.modifyPermissionArrayField.
If the same permission is granted to two users and the grantedTo property of that permission is updated at the same time both users will read the grantedTo property, the first will correctly update grantedTo, the second will overwrite (revert) the update of the first.

PermsAPI.modifyPermissionArrayField needs to be rewritten to become atomic. A single SQL UPDATE so that we don't have a time slot between a SELECT and then an UPDATE.

```
UPDATE diku_mod_permissions.permissions
SET jsonb = jsonb_set(jsonb, '{grantedTo}', jsonb->'grantedTo' || to_jsonb('33c5ef46-3586-488c-987d-e84116b8eff0'::text))
WHERE id='63351e09-f34c-4aac-a889-c6e8cd7f134e' AND NOT jsonb->'grantedTo' ? '33c5ef46-3586-488c-987d-e84116b8eff0';
```

```
UPDATE diku_mod_permissions.permissions
SET jsonb = jsonb_set(jsonb, '{grantedTo}', (jsonb->'grantedTo') - '33c5ef46-3586-488c-987d-e84116b8eff0'::text)
WHERE id='63351e09-f34c-4aac-a889-c6e8cd7f134e' AND jsonb->'grantedTo' ? '33c5ef46-3586-488c-987d-e84116b8eff0'::text;
```

Reply:
Ok. This could have been an issue before CompositeFuture.. Not because it's parallel in the CPU sense, but because events are interleaved.. Eg something can happen between "compose" calls. I have no clue on the SQL but it seems to work. Please check my code carefully. Thx.